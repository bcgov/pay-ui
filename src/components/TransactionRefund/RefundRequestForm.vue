<template>
  <div>
    <v-card flat>
      <v-card-title class="card-title">
        <v-icon
          class="pr-3"
          color="link"
        >
          mdi-file-document
        </v-icon>
        Refund Request
      </v-card-title>
      <v-card-text
        class="mt-5"
      >
        <v-alert v-if="previousRefundedAmount > 0"
          class="mt-4"
          border="left"
          dense
          outlined
          prominent
          shaped
          text
          :type="'error'"
        >
          This invoice has already been refunded for an amount of {{ formatAmount(previousRefundedAmount) }} previously.
        </v-alert>
        <v-form
          ref="refundRequestForm"
          v-model="isFormValid"
        >
          <v-row no-gutters>
            <v-col class="col-12 col-sm-12 ">
              <v-row>
                <v-col class="col-6 col-sm-3 font-weight-bold">
                  Refund Type
                </v-col>
                <v-col class="col-6 col-sm-9">
                  <v-radio-group
                    v-model="refundFormData.refundType"
                    :rules="formRules.refundType"
                    column
                    class="ma-0"
                    :hint="refundTypeHint"
                    persistent-hint
                  >
                    <v-radio
                      label="Full Refund"
                      :value="RefundType.FULL_REFUND"
                      :disabled="formDisabled"
                      @change="() => onRefundTypeChange()"
                    ></v-radio>
                    <v-radio
                      label="Partial Refund"
                      :value="RefundType.PARTIAL_REFUND"
                      :disabled="formDisabled || !isPartialRefundAllowed"
                      @change="() => onRefundTypeChange()"
                    ></v-radio>
                  </v-radio-group>
                </v-col>
              </v-row>
              <v-row v-if="refundFormData.refundType === RefundType.PARTIAL_REFUND">
                <v-col class="col-6 col-sm-3 font-weight-bold">
                  Specify Refund Amount
                </v-col>
                <v-col class="col-6 col-sm-9">
                  <div v-for="(lineItem, index) in refundFormData.refundLineItems" :key="lineItem.id">
                    <div class="mb-2">
                      {{ lineItem.description }} ({{ formatAmount(Number(lineItem.total)) }})
                    </div>
                    <v-checkbox
                      v-model="lineItem.refundEntireItemRequested"
                      @change="(value) => onRefundEntireItemRequestedChange(value, index)"
                      label="Refund entire item"
                    ></v-checkbox>
                    <div v-if="lineItem.filingFees" class="d-flex flex-wrap ga-4">
                      <div class="inline-input-box">
                        <v-text-field
                          v-model.number="lineItem.filingFeesRequested"
                          filled
                          :rules="getRequestedAmountRules(lineItem.filingFees)"
                          :suffix="`/${formatAmount(Number(lineItem.filingFees))}`"
                          :disabled="lineItem.refundEntireItemRequested"
                          label="Filing Fees"
                          type="number"
                          hide-spin-buttons
                          @change="() => calculateTotalRequestedAmount()"
                        />
                      </div>
                      <div v-if="lineItem.serviceFees" class="inline-input-box">
                        <v-text-field
                          v-model.number="lineItem.serviceFeesRequested"
                          filled
                          :rules="getRequestedAmountRules(lineItem.serviceFees)"
                          :suffix="`/${formatAmount(Number(lineItem.serviceFees))}`"
                          :disabled="lineItem.refundEntireItemRequested || true"
                          label="Service Fees"
                          type="number"
                          hide-spin-buttons
                          @change="() => calculateTotalRequestedAmount()"
                        />
                      </div>
                      <div v-if="lineItem.priorityFees" class="inline-input-box">
                        <v-text-field
                          v-model.number="lineItem.priorityFeesRequested"
                          filled
                          :rules="getRequestedAmountRules(lineItem.priorityFees)"
                          :suffix="`/${formatAmount(Number(lineItem.priorityFees))}`"
                          :disabled="lineItem.refundEntireItemRequested"
                          label="Priority Fees"
                          type="number"
                          hide-spin-buttons
                          @change="() => calculateTotalRequestedAmount()"
                        />
                      </div>
                      <div v-if="lineItem.futureEffectiveFees" class="inline-input-box">
                        <v-text-field
                          v-model.number="lineItem.futureEffectiveFeesRequested"
                          filled
                          :rules="getRequestedAmountRules(lineItem.futureEffectiveFees)"
                          :suffix="`/${formatAmount(Number(lineItem.futureEffectiveFees))}`"
                          :disabled="lineItem.refundEntireItemRequested"
                          label="Future Effective Fees"
                          type="number"
                          hide-spin-buttons
                          @change="() => calculateTotalRequestedAmount()"
                        />
                      </div>
                    </div>
                  </div>
                </v-col>
              </v-row>
              <v-row>
                <v-col class="col-6 col-sm-3 font-weight-bold">
                  Total Refund Amount
                </v-col>
                <v-col class="col-6 col-sm-9">
                  {{ formatAmount(Number(refundFormData.totalRefundAmount)) }}
                </v-col>
              </v-row>
              <v-row>
                <v-col class="col-6 col-sm-3 font-weight-bold">
                  Refund Method
                </v-col>
                <v-col v-if="refundMethods?.length == 1" class="col-6 col-sm-9">
                  {{ getRefundMethodText(refundMethods, refundFormData.refundMethod) }}
                </v-col>
                <v-col v-else class="col-6 col-sm-9">
                  <v-select
                    :items="refundMethods"
                    filled
                    required
                    v-model="refundFormData.refundMethod"
                    :rules="formRules.refundMethod"
                    :disabled="formDisabled"
                    placeholder="Select a Refund Method"
                  ></v-select>
                </v-col>
              </v-row>
              <v-row>
                <v-col class="col-6 col-sm-3 font-weight-bold">
                  Notification Email
                </v-col>
                <v-col class="col-6 col-sm-9">
                  <v-text-field
                    filled
                    required
                    v-model="refundFormData.notificationEmail"
                    :rules="formRules.notificationEmail"
                    :disabled="formDisabled"
                    maxlength="100"
                    type="email"
                    hint="Defaults to account email. Enter an email if missing, or to use a different address"
                    persistent-hint
                  />
                </v-col>
              </v-row>
              <v-row>
                <v-col class="col-6 col-sm-3 font-weight-bold">
                  Reasons for Refund
                </v-col>
                <v-col class="col-6 col-sm-9">
                  <v-text-field
                    filled
                    required
                    v-model="refundFormData.reasonsForRefund"
                    :rules="formRules.reasonsForRefund"
                    :disabled="formDisabled"
                    maxlength="250"
                    placeholder="Enter Reasons for Refund"
                    hint="The reason will be displayed in client's notification email when refund is approved"
                    persistent-hint
                  />
                </v-col>
              </v-row>
              <v-row>
                <v-col class="col-6 col-sm-3 font-weight-bold">
                  Staff Comment
                </v-col>
                <v-col class="col-6 col-sm-9">
                  <v-text-field
                    filled
                    required
                    v-model="refundFormData.staffComment"
                    :disabled="formDisabled"
                    maxlength="250"
                    placeholder="Enter Staff Comment (Optional)"
                    hint="Only viewable by staff"
                    persistent-hint
                  />
                </v-col>
              </v-row>
            </v-col>
          </v-row>
        </v-form>
      </v-card-text>
    </v-card>
    <div class="d-flex justify-space-between flex-wrap">
      <v-btn
        large
        outlined
        color="primary"
        class="mt-10"
      >
        <span>Cancel</span>
      </v-btn>
      <v-btn
        large
        @click="onReviewBtnClick()"
        :disabled="formDisabled || refundFormData.totalRefundAmount <= 0"
        color="primary"
        class="mt-10"
      >
        <span>Review and Confirm</span>
      </v-btn>
    </div>
  </div>
</template>

<script lang="ts">
import { defineComponent, reactive, toRefs, watch, PropType, nextTick, computed } from '@vue/composition-api'
import CommonUtils from '@/util/common-util'
import { RefundFormData, RefundType, RefundLineItem } from '@/models/transaction-refund'
import moment from 'moment'

export default defineComponent({
  name: 'RefundRequestForm',
  props: {
    totalTransactionAmount: {
      type: Number,
      required: true
    },
    refundLineItems: {
      type: Array as PropType<RefundLineItem[]>,
      required: true,
      default: () => []
    },
    refundMethods: {
      type: Array as PropType<{ text: string; value: string }[]>,
      required: true,
      default: () => []
    },
    previousRefundedAmount: {
      type: Number,
      required: true
    },
    isPartialRefundAllowed: {
      type: Boolean,
      required: false
    },
    invoicePaymentMethod: {
      type: String,
      required: false
    }
  },
  setup (props, { emit }) {
    const state = reactive({
      refundRequestForm: null,
      refundFormData: {
        refundType: null,
        refundLineItems: [],
        totalRefundAmount: 0,
        refundMethod: null,
        notificationEmail: null,
        reasonsForRefund: null,
        staffComment: null,
        requestedBy: null,
        requestedTime: null
      } as RefundFormData,
      isFormValid: false,
      formRules: {
        refundType: [
          (v: any) => !!v || 'Refund Type is required'
        ],
        refundMethod: [
          (v: any) => !!v || 'Refund Method is required'
        ],
        notificationEmail: CommonUtils.emailRules(true),
        reasonsForRefund: [
          (v: any) => !!v || 'Reasons for Refund is required'
        ]
      },
      formDisabled: computed(() => {
        return props.previousRefundedAmount > 0
      }),
      refundTypeHint: computed(() => {
        return props.isPartialRefundAllowed
          ? ''
          : `Partial Refunds not supported for payment method ${props.invoicePaymentMethod} invoices`
      })
    })

    const getRequestedAmountRules = (max?: number) => [
      (v: any) => v == null || v === '' || !isNaN(v) ? true : 'Enter a valid number',
      (v: any) => v == null || v === '' || Number(v) >= 0 ? true : 'Refund amount cannot be negative',
      (v: any) => v == null || v === '' || max == null || Number(v) <= Number(max)
        ? true
        : `Refund amount exceeds ${CommonUtils.formatAmount(Number(max))}`
    ]

    const currentUser = CommonUtils.getUserInfo()

    watch(() => props.refundLineItems, (newData) => {
      state.refundFormData.refundLineItems = JSON.parse(JSON.stringify(newData))
    }, { immediate: true, deep: true })

    watch(() => props.refundMethods, (newData) => {
      if (newData?.length === 1) {
        state.refundFormData.refundMethod = newData[0].value
      }
    }, { immediate: true })

    function onReviewBtnClick () {
      if (!state.refundRequestForm.validate()) {
        return
      }
      calculateTotalRequestedAmount()

      state.refundFormData.requestedBy = currentUser?.fullName
      state.refundFormData.requestedTime = moment().format()
      emit('onProceedToReview', state.refundFormData)
    }

    function onRefundEntireItemRequestedChange (value: boolean, index: number) {
      if (value) {
        const lineItem = state.refundFormData.refundLineItems[index]
        lineItem.filingFeesRequested = CommonUtils.formatToTwoDecimals(Number(lineItem.filingFees))
        lineItem.serviceFeesRequested = null
        lineItem.priorityFeesRequested = CommonUtils.formatToTwoDecimals(Number(lineItem.priorityFees))
        lineItem.futureEffectiveFeesRequested = CommonUtils.formatToTwoDecimals(Number(lineItem.futureEffectiveFees))
      }
      calculateTotalRequestedAmount()
    }

    function getValidNumber (amount: number, maxAmount: number) {
      return amount < 0 || (maxAmount != null && amount > maxAmount) ? 0 : amount
    }

    function calculateTotalRequestedAmount () {
      let total = 0
      if (state.refundFormData.refundType === RefundType.FULL_REFUND) {
        total = props.totalTransactionAmount
      } else {
        state.refundFormData.refundLineItems.forEach(item => {
          total += getValidNumber(Number(item.filingFeesRequested) || 0, item.filingFees)
          total += getValidNumber(Number(item.serviceFeesRequested) || 0, item.serviceFees)
          total += getValidNumber(Number(item.priorityFeesRequested) || 0, item.priorityFees)
          total += getValidNumber(Number(item.futureEffectiveFeesRequested) || 0, item.futureEffectiveFees)
        })
      }
      state.refundFormData.totalRefundAmount = total
    }

    async function onRefundTypeChange () {
      await nextTick()
      calculateTotalRequestedAmount()
    }

    return {
      ...toRefs(state),
      RefundType,
      getRequestedAmountRules,
      onReviewBtnClick,
      onRefundEntireItemRequestedChange,
      calculateTotalRequestedAmount,
      onRefundTypeChange,
      formatAmount: CommonUtils.formatAmount,
      getRefundMethodText: CommonUtils.getRefundMethodText
    }
  }
})
</script>

<style lang="scss" scoped>
@import '$assets/scss/theme.scss';
@import '$assets/scss/actions.scss';

.card-title {
  background-color: $app-lt-blue;
  justify-content: left;
  height: 75px;
  font-weight: bold;
  font-size: 1.125rem;
}
.ga-4 {
  gap: 16px;
}

.inline-input-box {
  width: 212px;
}

::v-deep {
  .v-sheet--shaped .v-alert__border--left {
    border-width: 0 !important;
  }
}
</style>
