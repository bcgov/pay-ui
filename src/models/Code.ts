// To handle codes tables such as SuspensionReasonCode, etc
export interface Code {
    code?: string;
    description?: string;
}

export interface PaymentMethod {
    code: string;
    description: string;
    partialRefund?: boolean;
    refundableStatuses?: string;
}
