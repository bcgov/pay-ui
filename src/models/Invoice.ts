// Model for storing invoice of a routing slip
export interface LineItem {
  description?: string
  filingFees?: number
  futureEffectiveFees?: number
  gst?: number
  id?: number
  priorityFees?: number
  pst?: number
  quantity?: number
  serviceFees?: number
  statusCode?: string
  total?: number
  waivedBy?: string
  waivedFees?: number
}

export interface Reference {
  id?: number
  invoiceNumber?: string
  statusCode?: string
}

export interface Detail {
  label?: string
  value?: string
}

export interface Invoice {
  businessIdentifier?: string
  corpTypeCode?: string
  createdBy?: string
  createdOn?: string
  createdName?: string
  id?: number
  lineItems?: LineItem[]
  detail?: Detail[]
  paid?: number,
  paymentAccount?: {
    accountId?: string
    accountName?: string
    billable?: boolean
  }
  folioNumber?: string
  paymentMethod?: string
  references?: Reference[]
  refund?: number
  routingSlip?: string
  serviceFees?: string
  statusCode?: string
  total?: number,
  details?: Detail[]
}

export interface InvoiceList {
  consInvNumber?: string
  invoiceNumber: string
  invoices: Invoice[],
  paymentMethod: string
  paymentSystem: string
  statusCode: string
  invoiceAmount: number
  paidAmount: number
}

export interface InvoiceListResponse {
  items: InvoiceList[]
  limit: number
  page: number
  total: number
}

// For displaying transaction data table
export interface InvoiceDisplay {
  createdOn?: string
  invoiceNumber?: string
  statusCode?: string
  total?: number
  createdName?: string
  createdBy?: string
  description?: string[]
  id?: number
}
