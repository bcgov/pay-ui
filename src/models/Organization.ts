import { Contact } from './contact'

export interface Organization {
  id?: number
}

export interface Organizations{
  orgs : Organization[]
}

export interface PADInfo {
  bankAccountNumber: string
  bankInstitutionNumber: string
  bankTransitNumber: string
  isTOSAccepted?: boolean
  isAcknowledged?: boolean
}

export interface CFSAccountDetails {
  bankAccountNumber: string
  bankInstitutionNumber: string
  bankTransitNumber: string
  cfsAccountNumber: string
  cfsPartyNumber: string
  cfsSiteNumber: string
  status: string
}

export interface GLInfo {
  client: string
  responsibilityCentre: string
  serviceLine: string
  stob: string
  projectCode: string
}

export interface OrgPaymentDetails {
  accountId: string
  accountName: string
  bcolAccount: string
  bcolUserId: string
  cfsAccount: CFSAccountDetails
  obCredit: string
  padCredit: string
  paymentMethod: string
  statementNotificationEnabled: true
  padTosAcceptedBy:string
  padTosAcceptedDate:string
  futurePaymentMethod: string
  revenueAccount?: GLInfo
  eftEnable?: boolean
}

export interface RoleInfo {
  icon: string
  name: string
  desc: string
  displayName: string
  displayOrder: number
  label: string
}

export interface AddUserBody {
  username: string
  password: string
  selectedRole?: RoleInfo
  membershipType: string
}

export interface AddUsersToOrgBody {
    users: AddUserBody[]
    orgId: string
}

export interface BulkUserResponse {
  httpStatus: number
  username: string
  created?: string
  error?: string
  firstname?: string
  type?: string
}

export interface BulkUserResponseBody {
  users: BulkUserResponse[]
}

export enum MembershipStatus {
  'Active' = 'ACTIVE',
  'Inactive' = 'INACTIVE',
  'Rejected' = 'REJECTED',
  'Pending' = 'PENDING_APPROVAL',
  'PendingStaffReview'='PENDING_STAFF_REVIEW'
}

export enum MembershipType {
  'Admin' = 'ADMIN',
  'Coordinator' = 'COORDINATOR',
  'User' = 'USER',
  'Staff' = 'STAFF',
  'ExternalStaff' = 'EXTERNAL_STAFF'
}

export interface UserTerms {
    isTermsOfUseAccepted: boolean
    termsOfUseAcceptedVersion: string
}

export interface User {
    firstname: string;
    lastname: string;
    username?: string;
    modified?: Date
    userTerms?: UserTerms
    contacts?: Contact[]
    email?: string,
    loginSource?: string
    id?: number
    keycloakGuid?: string
    verified?:boolean
}

export interface Member {
  id: number
  membershipTypeCode: MembershipType
  membershipStatus: MembershipStatus
  roleDisplayName?:string
  user: User
}
