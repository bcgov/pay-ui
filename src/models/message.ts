export interface Message {
  text: string
  type: 'success' | 'error'
  show: boolean
}
