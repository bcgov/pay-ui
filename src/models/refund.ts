export interface RefundRevenueType {
  paymentLineItemId: string
  refundAmount: number
  refundType: string
}

export interface RefundRequest {
  reason: string
  refundRevenue?: RefundRevenueType[]
}

export interface EftRefundBaseRequest {
  shortNameId: number
  refundAmount: number
  refundEmail: string
  comment?: string
}

export interface EftRefundRequest extends EftRefundBaseRequest{
  casSupplierNum: string
  casSupplierSite: string
}

export interface EftRefundChequeRequest extends EftRefundBaseRequest{
  entityName: string
  addressCity: string
  addressCountry: string
  addressRegion: string
  deliveryInstructions?: string
  postalCode: string
  streetAddress: string
  streetAddressAdditional?: string
}
