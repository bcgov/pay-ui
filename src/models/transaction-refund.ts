import { Detail } from './Invoice'

export interface PaymentData {
  accountName: string | null
  folioNumber: string | null
  initiatedBy: string | null
  paymentMethod: string | null
  paymentStatus: string | null
  totalTransactionAmount: number
}

export interface TransactionData {
  invoiceId: number | null
  transactionDate: string | null
  invoiceReferenceId: string | null
  invoiceStatusCode: string | null
  invoiceCreatedOn: string | null
  transactionAmount: number
  applicationName: string | null
  applicationType: string | null
  businessIdentifier: string | null
  applicationDetails: Detail[] | null
}

export enum RefundType {
  FULL_REFUND = 'Full Refund',
  PARTIAL_REFUND = 'Partial Refund'
}

export interface RefundLineItem {
  id: number | null
  description: string | null
  filingFees?: number | null
  serviceFees?: number | null
  priorityFees?: number | null
  total?: number | null
  futureEffectiveFees?: number | null
  filingFeesRequested?: number | string | null
  serviceFeesRequested?: number | string | null
  priorityFeesRequested?: number | string | null
  futureEffectiveFeesRequested?: number | string | null
  refundEntireItemRequested?: boolean | null
}

export interface RefundFormData {
  refundType: RefundType | null
  refundLineItems: RefundLineItem[] | null
  totalRefundAmount: number | null
  refundMethod: string | null
  notificationEmail: string | null
  reasonsForRefund: string | null
  staffComment: string | null
  requestedBy: string | null
  requestedTime: string | null
}

export enum RefundRequestStage {
  REQUEST_FORM = 'Request Form',
  DATA_VALIDATED = 'DATA Validated'
}
