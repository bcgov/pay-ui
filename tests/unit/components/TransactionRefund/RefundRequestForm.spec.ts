import { createLocalVue, shallowMount } from '@vue/test-utils'
import Vuetify from 'vuetify'

import RefundRequestForm from '@/components/TransactionRefund/RefundRequestForm.vue'
import CommonUtils from '@/util/common-util'
import { RefundType } from '@/models/transaction-refund'

describe('RefundRequestForm.vue', () => {
  const localVue = createLocalVue()
  const vuetify = new Vuetify({})

  beforeEach(() => {
    vi.resetModules()
    vi.clearAllMocks()
  })

  it('renders component', () => {
    const wrapper = shallowMount(RefundRequestForm, {
      localVue,
      vuetify,
      propsData: {
        totalTransactionAmount: 2000,
        refundLineItems: [],
        refundMethods: [],
        previousRefundedAmount: 0,
        isPartialRefundAllowed: true
      },
      directives: { can () { } }
    })

    expect(wrapper.findComponent(RefundRequestForm).exists()).toBe(true)
    expect(wrapper.text()).toContain('Refund Request')
    expect(wrapper.text()).toContain('Refund Type')
    expect(wrapper.text()).toContain('Total Refund Amount')
    expect(wrapper.text()).toContain('Refund Method')
    expect(wrapper.text()).toContain('Notification Email')
    expect(wrapper.text()).toContain('Reasons for Refund')
    expect(wrapper.text()).toContain('Staff Comment')
  })

  it('displays data values', async () => {
    const wrapper: any = shallowMount(RefundRequestForm, {
      localVue,
      vuetify,
      propsData: {
        totalTransactionAmount: 2000,
        refundLineItems: [],
        refundMethods: [{ text: 'Credit Card', value: 'credit-card' }],
        previousRefundedAmount: 0,
        isPartialRefundAllowed: true
      },
      directives: { can () { } }
    })

    wrapper.vm.refundFormData.refundType = RefundType.FULL_REFUND
    await wrapper.vm.onRefundTypeChange()
    wrapper.vm.refundFormData.refundMethod = 'credit-card'
    wrapper.vm.refundFormData.notificationEmail = 'test@example.com'
    wrapper.vm.refundFormData.reasonsForRefund = 'Reason'
    wrapper.vm.refundFormData.staffComment = 'Comment'
    await wrapper.vm.$nextTick()

    expect(wrapper.text()).toContain(CommonUtils.formatAmount(2000))
    expect(wrapper.text()).toContain('Credit Card')
    expect(wrapper.vm.refundFormData.notificationEmail).toBe('test@example.com')
    expect(wrapper.vm.refundFormData.reasonsForRefund).toBe('Reason')
    expect(wrapper.vm.refundFormData.staffComment).toBe('Comment')
  })

  it('emits onProceedToReview on valid submit', async () => {
    const wrapper: any = shallowMount(RefundRequestForm, {
      localVue,
      vuetify,
      propsData: {
        totalTransactionAmount: 2000,
        refundLineItems: [],
        refundMethods: [{ text: 'Credit Card', value: 'credit-card' }],
        previousRefundedAmount: 0,
        isPartialRefundAllowed: true
      },
      directives: { can () { } }
    })

    wrapper.vm.refundFormData.refundType = RefundType.FULL_REFUND
    wrapper.vm.refundFormData.refundMethod = 'credit-card'
    wrapper.vm.refundFormData.notificationEmail = 'test@example.com'
    wrapper.vm.refundFormData.reasonsForRefund = 'Reason'
    wrapper.vm.refundFormData.staffComment = 'Comment'

    wrapper.vm.refundRequestForm = { validate: () => true }

    wrapper.vm.onReviewBtnClick()
    await wrapper.vm.$nextTick()

    expect(wrapper.emitted('onProceedToReview')).toBeTruthy()
    const payload = wrapper.emitted('onProceedToReview')[0][0]
    expect(payload.refundType).toContain(RefundType.FULL_REFUND)
  })

  it('matches snapshot', () => {
    const wrapper = shallowMount(RefundRequestForm, {
      localVue,
      vuetify,
      propsData: { totalTransactionAmount: 2000 },
      directives: { can () { } }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})
