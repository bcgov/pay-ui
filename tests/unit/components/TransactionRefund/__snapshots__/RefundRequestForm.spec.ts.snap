// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RefundRequestForm.vue > matches snapshot 1`] = `
"<div data-v-6eb9f862="">
  <v-card-stub data-v-6eb9f862="" loaderheight="4" tag="div" flat="true">
    <v-card-title-stub data-v-6eb9f862="" tag="div" class="card-title">
      <v-icon-stub data-v-6eb9f862="" color="link" class="pr-3"> mdi-file-document </v-icon-stub> Refund Request
    </v-card-title-stub>
    <v-card-text-stub data-v-6eb9f862="" tag="div" class="mt-5">
      <!---->
      <v-form-stub data-v-6eb9f862="">
        <v-row-stub data-v-6eb9f862="" tag="div" nogutters="true">
          <v-col-stub data-v-6eb9f862="" tag="div" class="col-12 col-sm-12 ">
            <v-row-stub data-v-6eb9f862="" tag="div">
              <v-col-stub data-v-6eb9f862="" tag="div" class="col-6 col-sm-3 font-weight-bold"> Refund Type </v-col-stub>
              <v-col-stub data-v-6eb9f862="" tag="div" class="col-6 col-sm-9">
                <v-radio-group-stub data-v-6eb9f862="" valuecomparator="[Function]" activeclass="v-item--active" tag="div" errorcount="1" errormessages="" messages="" rules="[Function]" successmessages="" backgroundcolor="" height="auto" hint="Partial Refunds not supported for payment method undefined invoices" persistenthint="true" column="true" class="ma-0">
                  <v-radio-stub data-v-6eb9f862="" ripple="true" label="Full Refund" officon="$radioOff" onicon="$radioOn" value="Full Refund"></v-radio-stub>
                  <v-radio-stub data-v-6eb9f862="" ripple="true" disabled="true" label="Partial Refund" officon="$radioOff" onicon="$radioOn" value="Partial Refund"></v-radio-stub>
                </v-radio-group-stub>
              </v-col-stub>
            </v-row-stub>
            <!---->
            <v-row-stub data-v-6eb9f862="" tag="div">
              <v-col-stub data-v-6eb9f862="" tag="div" class="col-6 col-sm-3 font-weight-bold"> Total Refund Amount </v-col-stub>
              <v-col-stub data-v-6eb9f862="" tag="div" class="col-6 col-sm-9"> $0.00 </v-col-stub>
            </v-row-stub>
            <v-row-stub data-v-6eb9f862="" tag="div">
              <v-col-stub data-v-6eb9f862="" tag="div" class="col-6 col-sm-3 font-weight-bold"> Refund Method </v-col-stub>
              <v-col-stub data-v-6eb9f862="" tag="div" class="col-6 col-sm-9">
                <v-select-stub data-v-6eb9f862="" errorcount="1" errormessages="" messages="" rules="[Function]" successmessages="" appendicon="$dropdown" backgroundcolor="" loaderheight="2" clearicon="$clear" filled="true" placeholder="Select a Refund Method" type="text" valuecomparator="[Function]" nodatatext="$vuetify.noDataText" items="" itemcolor="primary" itemdisabled="disabled" itemtext="text" itemvalue="value" menuprops="[object Object]" required=""></v-select-stub>
              </v-col-stub>
            </v-row-stub>
            <v-row-stub data-v-6eb9f862="" tag="div">
              <v-col-stub data-v-6eb9f862="" tag="div" class="col-6 col-sm-3 font-weight-bold"> Notification Email </v-col-stub>
              <v-col-stub data-v-6eb9f862="" tag="div" class="col-6 col-sm-9">
                <v-text-field-stub data-v-6eb9f862="" errorcount="1" errormessages="" messages="" rules="[Function]" successmessages="" backgroundcolor="" hint="Defaults to account email. Enter an email if missing, or to use a different address" persistenthint="true" loaderheight="2" clearicon="$clear" filled="true" type="email" required=""></v-text-field-stub>
              </v-col-stub>
            </v-row-stub>
            <v-row-stub data-v-6eb9f862="" tag="div">
              <v-col-stub data-v-6eb9f862="" tag="div" class="col-6 col-sm-3 font-weight-bold"> Reasons for Refund </v-col-stub>
              <v-col-stub data-v-6eb9f862="" tag="div" class="col-6 col-sm-9">
                <v-text-field-stub data-v-6eb9f862="" errorcount="1" errormessages="" messages="" rules="[Function]" successmessages="" backgroundcolor="" hint="The reason will be displayed in client's notification email when refund is approved" persistenthint="true" loaderheight="2" clearicon="$clear" filled="true" placeholder="Enter Reasons for Refund" type="text" required=""></v-text-field-stub>
              </v-col-stub>
            </v-row-stub>
            <v-row-stub data-v-6eb9f862="" tag="div">
              <v-col-stub data-v-6eb9f862="" tag="div" class="col-6 col-sm-3 font-weight-bold"> Staff Comment </v-col-stub>
              <v-col-stub data-v-6eb9f862="" tag="div" class="col-6 col-sm-9">
                <v-text-field-stub data-v-6eb9f862="" errorcount="1" errormessages="" messages="" rules="" successmessages="" backgroundcolor="" hint="Only viewable by staff" persistenthint="true" loaderheight="2" clearicon="$clear" filled="true" placeholder="Enter Staff Comment (Optional)" type="text" required=""></v-text-field-stub>
              </v-col-stub>
            </v-row-stub>
          </v-col-stub>
        </v-row-stub>
      </v-form-stub>
    </v-card-text-stub>
  </v-card-stub>
  <div data-v-6eb9f862="" class="d-flex justify-space-between flex-wrap">
    <v-btn-stub data-v-6eb9f862="" color="primary" outlined="true" tag="button" activeclass="" large="true" type="button" class="mt-10"><span data-v-6eb9f862="">Cancel</span></v-btn-stub>
    <v-btn-stub data-v-6eb9f862="" color="primary" tag="button" activeclass="" disabled="true" large="true" type="button" class="mt-10"><span data-v-6eb9f862="">Review and Confirm</span></v-btn-stub>
  </div>
</div>"
`;
